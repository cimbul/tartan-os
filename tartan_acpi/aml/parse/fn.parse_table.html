<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `parse_table` fn in crate `tartan_acpi`."><meta name="keywords" content="rust, rustlang, rust-lang, parse_table"><title>tartan_acpi::aml::parse::parse_table - Rust</title><link rel="stylesheet" type="text/css" href="../../../normalize.css"><link rel="stylesheet" type="text/css" href="../../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../../../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../../../ayu.css" disabled ><script id="default-settings"></script><script src="../../../storage.js"></script><noscript><link rel="stylesheet" href="../../../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../../../favicon.svg">
<link rel="alternate icon" type="image/png" href="../../../favicon-16x16.png">
<link rel="alternate icon" type="image/png" href="../../../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../../../down-arrow.svg");}</style></head><body class="rustdoc fn"><!--[if lte IE 8]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu">&#9776;</div><a href='../../../tartan_acpi/index.html'><div class='logo-container rust-logo'><img src='../../../rust-logo.png' alt='logo'></div></a><div class="sidebar-elems"><p class="location"><a href="../../index.html">tartan_acpi</a>::<wbr><a href="../index.html">aml</a>::<wbr><a href="index.html">parse</a></p><div id="sidebar-vars" data-name="parse_table" data-ty="fn" data-relpath=""></div><script defer src="sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu"><img src="../../../brush.svg" width="18" alt="Pick another theme!"></button><div id="theme-choices" role="menu"></div></div><script src="../../../theme.js"></script><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" class="help-button">?</button>
                <a id="settings-menu" href="../../../settings.html"><img src="../../../wheel.svg" width="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Function <a href="../../index.html">tartan_acpi</a>::<wbr><a href="../index.html">aml</a>::<wbr><a href="index.html">parse</a>::<wbr><a class="fn" href="">parse_table</a></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../../../src/tartan_acpi/aml/parse.rs.html#42-75" title="goto source code">[src]</a></span></h1><pre class="rust fn">pub fn parse_table(<br>&nbsp;&nbsp;&nbsp;&nbsp;data: <a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.slice.html">&amp;[</a><a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.u8.html">u8</a><a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.slice.html">]</a><br>) -&gt; <a class="enum" href="https://doc.rust-lang.org/nightly/core/result/enum.Result.html" title="enum core::result::Result">Result</a>&lt;<a class="struct" href="../../../tartan_acpi/aml/struct.AMLTable.html" title="struct tartan_acpi::aml::AMLTable">AMLTable</a>&lt;'_&gt;, <a class="struct" href="../../../tartan_parsers/error/struct.ErrorWithPosition.html" title="struct tartan_parsers::error::ErrorWithPosition">ErrorWithPosition</a>&lt;'_, <a class="struct" href="../../../tartan_acpi/aml/parse/state/struct.ParserState.html" title="struct tartan_acpi::aml::parse::state::ParserState">ParserState</a>&lt;'_&gt;&gt;&gt;</pre><div class="docblock"><p>Parse an ACPI table that contains AML, including the header</p>
<h1 id="errors" class="section-header"><a href="#errors">Errors</a></h1>
<p>Returns an error if the AML contents cannot be parsed.</p>
</div></section><section id="search" class="content hidden"></section><section class="footer"></section><div id="rustdoc-vars" data-root-path="../../../" data-current-crate="tartan_acpi"></div>
    <script src="../../../main.js"></script><script defer src="../../../search-index.js"></script></body></html>