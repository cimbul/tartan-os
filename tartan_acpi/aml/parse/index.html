<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Parsers for AML bytecode."><meta name="keywords" content="rust, rustlang, rust-lang, parse"><title>tartan_acpi::aml::parse - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" href="../../../normalize.css"><link rel="stylesheet" href="../../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" href="../../../ayu.css" disabled><link rel="stylesheet" href="../../../dark.css" disabled><link rel="stylesheet" href="../../../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../../../storage.js"></script><script defer src="../../../main.js"></script><noscript><link rel="stylesheet" href="../../../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../../../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../../../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../../../favicon.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../../../tartan_acpi/index.html"><div class="logo-container"><img class="rust-logo" src="../../../rust-logo.svg" alt="logo"></div></a><h2 class="location"></h2></nav><nav class="sidebar"><a class="sidebar-logo" href="../../../tartan_acpi/index.html"><div class="logo-container"><img class="rust-logo" src="../../../rust-logo.svg" alt="logo"></div></a><h2 class="location"><a href="#">Module parse</a></h2><div class="sidebar-elems"><section><div class="block"><ul><li><a href="#modules">Modules</a></li><li><a href="#traits">Traits</a></li><li><a href="#functions">Functions</a></li><li><a href="#types">Type Definitions</a></li></ul></div></section></div></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../../../tartan_acpi/index.html"><img class="rust-logo" src="../../../rust-logo.svg" alt="logo"></a><nav class="sub"><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><button type="button">?</button></div><div id="settings-menu" tabindex="-1"><a href="../../../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../../../wheel.svg"></a></div></div></form></nav></div><section id="main-content" class="content"><div class="main-heading"><h1 class="fqn"><span class="in-band">Module <a href="../../index.html">tartan_acpi</a>::<wbr><a href="../index.html">aml</a>::<wbr><a class="mod" href="#">parse</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../../../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span></h1><span class="out-of-band"><a class="srclink" href="../../../src/tartan_acpi/aml/parse.rs.html#1-2954">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Parsers for AML bytecode.</p>
<p>AML bytecode grammar is defined in §20.2. The top-level production is:</p>
<div class="example-wrap"><pre class="language-text"><code>AMLCode := DefBlockHeader TermList</code></pre></div>
<p>Everything in <code>DefBlockHeader</code> just repeats the fixed
<a href="../../struct.DescriptionHeader.html"><code>DescriptionHeader</code></a> struct used in other tables,
so there’s no reason to write a parser for it.</p>
</div></details><h2 id="modules" class="small-section-header"><a href="#modules">Modules</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="mod" href="data/index.html" title="tartan_acpi::aml::parse::data mod">data</a></div><div class="item-right docblock-short">Data objects, defined in §20.2.3</div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="misc/index.html" title="tartan_acpi::aml::parse::misc mod">misc</a></div><div class="item-right docblock-short">Miscellaneous objects, defined in §20.2.6</div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="name/index.html" title="tartan_acpi::aml::parse::name mod">name</a></div><div class="item-right docblock-short">Name objects, defined in §20.2.2</div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="state/index.html" title="tartan_acpi::aml::parse::state mod">state</a></div><div class="item-right docblock-short">Parser context needed to disambiguate grammar</div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="term/index.html" title="tartan_acpi::aml::parse::term mod">term</a></div><div class="item-right docblock-short">Terms, defined in §20.2.5</div></div></div><h2 id="traits" class="small-section-header"><a href="#traits">Traits</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.AMLParseError.html" title="tartan_acpi::aml::parse::AMLParseError trait">AMLParseError</a></div><div class="item-right docblock-short">Trait “alias” for error type used by AML parsers</div></div><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.Parse.html" title="tartan_acpi::aml::parse::Parse trait">Parse</a></div><div class="item-right docblock-short">An object that can be parsed from AML bytecode</div></div></div><h2 id="functions" class="small-section-header"><a href="#functions">Functions</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.parse_table.html" title="tartan_acpi::aml::parse::parse_table fn">parse_table</a></div><div class="item-right docblock-short">Parse an ACPI table that contains AML, including the header</div></div></div><h2 id="types" class="small-section-header"><a href="#types">Type Definitions</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="type" href="type.AMLParseResult.html" title="tartan_acpi::aml::parse::AMLParseResult type">AMLParseResult</a></div><div class="item-right docblock-short">Alias for type returned by AML parsers</div></div></div></section></div></main><div id="rustdoc-vars" data-root-path="../../../" data-current-crate="tartan_acpi" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.66.0-nightly (c97d02cdb 2022-10-05)" ></div></body></html>