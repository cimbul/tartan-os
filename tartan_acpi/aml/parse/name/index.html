<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Name objects, defined in §20.2.2"><title>tartan_acpi::aml::parse::name - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../../../static.files/rustdoc-b25d77b18a16a9a4.css"><meta name="rustdoc-vars" data-root-path="../../../../" data-static-root-path="../../../../static.files/" data-current-crate="tartan_acpi" data-themes="" data-resource-suffix="" data-rustdoc-version="1.82.0-nightly (506052d49 2024-08-16)" data-channel="nightly" data-search-js="search-ef54e092fea9c776.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../../../../static.files/storage-118b08c4c78b968e.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../../../static.files/main-cb0df477c2d67d00.js"></script><noscript><link rel="stylesheet" href="../../../../static.files/noscript-3b12f09e550e0385.css"></noscript><link rel="alternate icon" type="image/png" href="../../../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../../../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../../../tartan_acpi/index.html">tartan_<wbr>acpi</a><span class="version">0.1.0</span></h2></div><h2 class="location"><a href="#">Module name</a></h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#functions">Functions</a></li></ul></section><h2><a href="../index.html">In tartan_<wbr>acpi::<wbr>aml::<wbr>parse</a></h2></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Module <a href="../../../index.html">tartan_acpi</a>::<wbr><a href="../../index.html">aml</a>::<wbr><a href="../index.html">parse</a>::<wbr><a class="mod" href="#">name</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../../../../src/tartan_acpi/aml/parse.rs.html#422">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Name objects, defined in §20.2.2</p>
</div></details><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="fn" href="fn.parse_target.html" title="fn tartan_acpi::aml::parse::name::parse_target">parse_<wbr>target</a></div><div class="desc docblock-short">Grammar:</div></li><li><div class="item-name"><a class="fn" href="fn.super_name_minus_invoke.html" title="fn tartan_acpi::aml::parse::name::super_name_minus_invoke">super_<wbr>name_<wbr>minus_<wbr>invoke</a></div><div class="desc docblock-short">Parse a <a href="../../name/enum.SuperName.html" title="enum tartan_acpi::aml::name::SuperName"><code>SuperName</code></a>, but do <em>not</em> allow method invocations.</div></li></ul></section></div></main></body></html>