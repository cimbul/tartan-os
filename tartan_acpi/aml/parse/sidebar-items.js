initSidebarItems({"fn":[["parse_table","Parse an ACPI table that contains AML, including the header"]],"mod":[["data","Data objects, defined in §20.2.3"],["misc","Miscellaneous objects, defined in §20.2.6"],["name","Name objects, defined in §20.2.2"],["state","Parser context needed to disambiguate grammar"],["term","Terms, defined in §20.2.5"]],"trait":[["AMLParseError","Trait “alias” for error type used by AML parsers"],["Parse","An object that can be parsed from AML bytecode"]],"type":[["AMLParseResult","Alias for type returned by AML parsers"]]});