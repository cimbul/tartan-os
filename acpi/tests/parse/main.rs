#![warn(clippy::pedantic)]

use insta::{assert_debug_snapshot, glob, with_settings};
use std::fs;
use tartan_acpi::aml::parse::parse_table;

#[test]
fn parse_all() {
    // Parse all AML files and compare to past output
    glob!("aml/*.aml", |aml_path| {
        // Make the autogenerated snapshot filenames less wordy
        let test_name = aml_path.file_stem().unwrap().to_str().unwrap();
        with_settings!({snapshot_suffix => ""}, {
            let aml_data = fs::read(aml_path).unwrap();
            assert_debug_snapshot!(test_name, parse_table(&aml_data));
        });
    });
}
