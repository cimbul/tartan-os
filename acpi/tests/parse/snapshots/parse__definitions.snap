---
source: acpi/tests/parse/main.rs
expression: parse_table(&aml_data)
---
Ok(
    AMLTable {
        header: DescriptionHeader {
            signature: [
                83,
                83,
                68,
                84,
            ],
            length: 629,
            revision: 2,
            checksum: 50,
            oem_id: [
                84,
                97,
                114,
                116,
                97,
                110,
            ],
            oem_table_id: [
                68,
                97,
                116,
                97,
                0,
                0,
                0,
                0,
            ],
            oem_revision: 0,
            creator_id: 1280593481,
            creator_revision: 538969879,
        },
        body: [
            Modifier(
                Name(
                    NameString["U8__"],
                    Data(
                        Data(
                            Byte(
                                255,
                            ),
                        ),
                    ),
                ),
            ),
            Modifier(
                Name(
                    NameString["U16_"],
                    Data(
                        Data(
                            Word(
                                65535,
                            ),
                        ),
                    ),
                ),
            ),
            Modifier(
                Name(
                    NameString["U32_"],
                    Data(
                        Data(
                            DWord(
                                4294967295,
                            ),
                        ),
                    ),
                ),
            ),
            Modifier(
                Name(
                    NameString["U64_"],
                    Data(
                        Data(
                            QWord(
                                1152921504606846975,
                            ),
                        ),
                    ),
                ),
            ),
            Modifier(
                Name(
                    NameString["STR0"],
                    Data(
                        Data(
                            String(
                                "",
                            ),
                        ),
                    ),
                ),
            ),
            Modifier(
                Name(
                    NameString["STR_"],
                    Data(
                        Data(
                            String(
                                "Hello, world!",
                            ),
                        ),
                    ),
                ),
            ),
            Modifier(
                Name(
                    NameString["Z___"],
                    Data(
                        Data(
                            Zero,
                        ),
                    ),
                ),
            ),
            Modifier(
                Name(
                    NameString["O___"],
                    Data(
                        Data(
                            One,
                        ),
                    ),
                ),
            ),
            Modifier(
                Name(
                    NameString["F___"],
                    Data(
                        Data(
                            Ones,
                        ),
                    ),
                ),
            ),
            Modifier(
                Name(
                    NameString["BUF0"],
                    Data(
                        Data(
                            Buffer(
                                Buffer {
                                    size: Data(
                                        Data(
                                            Byte(
                                                0,
                                            ),
                                        ),
                                    ),
                                    initializer: [],
                                },
                            ),
                        ),
                    ),
                ),
            ),
            Modifier(
                Name(
                    NameString["BUFF"],
                    Data(
                        Data(
                            Buffer(
                                Buffer {
                                    size: Data(
                                        Data(
                                            Byte(
                                                16,
                                            ),
                                        ),
                                    ),
                                    initializer: [
                                        1,
                                        2,
                                        3,
                                    ],
                                },
                            ),
                        ),
                    ),
                ),
            ),
            Modifier(
                Name(
                    NameString["VBUF"],
                    Data(
                        Data(
                            Buffer(
                                Buffer {
                                    size: Name(
                                        NameString["\\U32_"],
                                    ),
                                    initializer: [],
                                },
                            ),
                        ),
                    ),
                ),
            ),
            Modifier(
                Name(
                    NameString["PKG0"],
                    Data(
                        Package(
                            Package {
                                count: 0,
                                initializers: [],
                            },
                        ),
                    ),
                ),
            ),
            Modifier(
                Name(
                    NameString["PKG1"],
                    Data(
                        Package(
                            Package {
                                count: 32,
                                initializers: [
                                    Ref(
                                        Data(
                                            Data(
                                                Byte(
                                                    255,
                                                ),
                                            ),
                                        ),
                                    ),
                                    Ref(
                                        Data(
                                            Data(
                                                String(
                                                    "foo",
                                                ),
                                            ),
                                        ),
                                    ),
                                    Ref(
                                        Data(
                                            Data(
                                                Ones,
                                            ),
                                        ),
                                    ),
                                    Ref(
                                        Data(
                                            Data(
                                                Buffer(
                                                    Buffer {
                                                        size: Data(
                                                            Data(
                                                                Byte(
                                                                    2,
                                                                ),
                                                            ),
                                                        ),
                                                        initializer: [
                                                            255,
                                                            238,
                                                        ],
                                                    },
                                                ),
                                            ),
                                        ),
                                    ),
                                    Ref(
                                        Data(
                                            Package(
                                                Package {
                                                    count: 2,
                                                    initializers: [
                                                        Ref(
                                                            Data(
                                                                Data(
                                                                    DWord(
                                                                        3405691582,
                                                                    ),
                                                                ),
                                                            ),
                                                        ),
                                                        Ref(
                                                            Data(
                                                                Data(
                                                                    DWord(
                                                                        3735928559,
                                                                    ),
                                                                ),
                                                            ),
                                                        ),
                                                    ],
                                                },
                                            ),
                                        ),
                                    ),
                                    Name(
                                        NameString["\\U32_"],
                                    ),
                                    Name(
                                        NameString["\\STR_"],
                                    ),
                                    Name(
                                        NameString["\\BUFF"],
                                    ),
                                    Name(
                                        NameString["\\FLD3"],
                                    ),
                                    Name(
                                        NameString["\\VPKG"],
                                    ),
                                    Name(
                                        NameString["\\DEV1"],
                                    ),
                                    Name(
                                        NameString["\\EVT1"],
                                    ),
                                    Name(
                                        NameString["\\MTH1"],
                                    ),
                                    Name(
                                        NameString["\\MTX1"],
                                    ),
                                    Name(
                                        NameString["\\REG1"],
                                    ),
                                    Name(
                                        NameString["\\PWR1"],
                                    ),
                                    Name(
                                        NameString["\\PRC1"],
                                    ),
                                    Name(
                                        NameString["\\THZ1"],
                                    ),
                                ],
                            },
                        ),
                    ),
                ),
            ),
            Modifier(
                Name(
                    NameString["VPKG"],
                    Data(
                        VarPackage(
                            VarPackage {
                                count: Name(
                                    NameString["\\U32_"],
                                ),
                                initializers: [
                                    Ref(
                                        Data(
                                            Data(
                                                String(
                                                    "stuff",
                                                ),
                                            ),
                                        ),
                                    ),
                                    Ref(
                                        Data(
                                            Data(
                                                Zero,
                                            ),
                                        ),
                                    ),
                                ],
                            },
                        ),
                    ),
                ),
            ),
            Modifier(
                Name(
                    NameString["VPKG.OTHR"],
                    Data(
                        VarPackage(
                            VarPackage {
                                count: Data(
                                    Data(
                                        Word(
                                            256,
                                        ),
                                    ),
                                ),
                                initializers: [
                                    Ref(
                                        Data(
                                            Data(
                                                String(
                                                    "other",
                                                ),
                                            ),
                                        ),
                                    ),
                                    Ref(
                                        Data(
                                            Data(
                                                Word(
                                                    43981,
                                                ),
                                            ),
                                        ),
                                    ),
                                ],
                            },
                        ),
                    ),
                ),
            ),
            Named(
                DataTableRegion {
                    name: NameString["REG1"],
                    signature: Data(
                        Data(
                            String(
                                "XYZ1",
                            ),
                        ),
                    ),
                    oem_id: Data(
                        Data(
                            String(
                                "Tartan",
                            ),
                        ),
                    ),
                    oem_table_id: Data(
                        Data(
                            String(
                                "SomeTable",
                            ),
                        ),
                    ),
                },
            ),
            Named(
                OperationRegion {
                    name: NameString["REG2"],
                    region_space: GenericSerialBus,
                    offset: Data(
                        Data(
                            Byte(
                                66,
                            ),
                        ),
                    ),
                    length: Data(
                        Data(
                            Byte(
                                48,
                            ),
                        ),
                    ),
                },
            ),
            Modifier(
                Name(
                    NameString["COM1"],
                    Data(
                        Data(
                            Buffer(
                                Buffer {
                                    size: Data(
                                        Data(
                                            Byte(
                                                27,
                                            ),
                                        ),
                                    ),
                                    initializer: [
                                        142,
                                        22,
                                        0,
                                        2,
                                        0,
                                        1,
                                        2,
                                        0,
                                        0,
                                        1,
                                        8,
                                        0,
                                        0,
                                        119,
                                        1,
                                        0,
                                        18,
                                        0,
                                        3,
                                        4,
                                        92,
                                        73,
                                        50,
                                        67,
                                        0,
                                        121,
                                        0,
                                    ],
                                },
                            ),
                        ),
                    ),
                ),
            ),
            Named(
                BankField {
                    region_name: NameString["REG2"],
                    bank_name: NameString["FLD6"],
                    bank_value: Data(
                        Data(
                            Word(
                                4660,
                            ),
                        ),
                    ),
                    flags: FieldFlags {
                        access_type: Buffer,
                        lock: true,
                        update_rule: Preserve,
                    },
                    elements: [
                        Named {
                            name: NameSeg["FLD1"],
                            bit_length: 10,
                        },
                        Named {
                            name: NameSeg["FLD2"],
                            bit_length: 4,
                        },
                        Reserved {
                            bit_length: 178,
                        },
                        ConnectNamed(
                            NameString["COM1"],
                        ),
                        AccessAs(
                            Buffer,
                            Bytes(
                                4,
                            ),
                        ),
                        Named {
                            name: NameSeg["FLD3"],
                            bit_length: 5,
                        },
                        ConnectBuffer(
                            Buffer {
                                size: Data(
                                    Data(
                                        Byte(
                                            25,
                                        ),
                                    ),
                                ),
                                initializer: [
                                    142,
                                    22,
                                    0,
                                    2,
                                    0,
                                    1,
                                    2,
                                    0,
                                    0,
                                    1,
                                    8,
                                    0,
                                    0,
                                    119,
                                    1,
                                    0,
                                    18,
                                    0,
                                    3,
                                    4,
                                    92,
                                    73,
                                    50,
                                    67,
                                    0,
                                ],
                            },
                        ),
                    ],
                },
            ),
            Named(
                Field {
                    region_name: NameString["REG1"],
                    flags: FieldFlags {
                        access_type: QWord,
                        lock: true,
                        update_rule: WriteAsZeros,
                    },
                    elements: [],
                },
            ),
            Named(
                IndexField {
                    index_name: NameString["FLD1"],
                    data_name: NameString["FLD2"],
                    flags: FieldFlags {
                        access_type: DWord,
                        lock: false,
                        update_rule: WriteAsOnes,
                    },
                    elements: [],
                },
            ),
            Named(
                CreateField {
                    source_buffer: Name(
                        NameString["BUFF"],
                    ),
                    bit_index: Data(
                        Data(
                            Byte(
                                3,
                            ),
                        ),
                    ),
                    num_bits: Data(
                        Data(
                            Byte(
                                5,
                            ),
                        ),
                    ),
                    name: NameString["FLD4"],
                },
            ),
            Named(
                CreateBitField {
                    source_buffer: Name(
                        NameString["BUFF"],
                    ),
                    bit_index: Data(
                        Data(
                            Byte(
                                9,
                            ),
                        ),
                    ),
                    name: NameString["FLD5"],
                },
            ),
            Named(
                CreateByteField {
                    source_buffer: Name(
                        NameString["BUFF"],
                    ),
                    byte_index: Data(
                        Data(
                            Byte(
                                2,
                            ),
                        ),
                    ),
                    name: NameString["FLD6"],
                },
            ),
            Named(
                CreateWordField {
                    source_buffer: Name(
                        NameString["BUFF"],
                    ),
                    byte_index: Data(
                        Data(
                            Byte(
                                8,
                            ),
                        ),
                    ),
                    name: NameString["FLD7"],
                },
            ),
            Named(
                CreateDWordField {
                    source_buffer: Name(
                        NameString["BUFF"],
                    ),
                    byte_index: Data(
                        Data(
                            Byte(
                                4,
                            ),
                        ),
                    ),
                    name: NameString["FLD8"],
                },
            ),
            Named(
                CreateQWordField {
                    source_buffer: Name(
                        NameString["BUFF"],
                    ),
                    byte_index: Data(
                        Data(
                            Byte(
                                7,
                            ),
                        ),
                    ),
                    name: NameString["FLD9"],
                },
            ),
            Named(
                Device {
                    name: NameString["DEV1"],
                    body: [],
                },
            ),
            Named(
                Event(
                    NameString["EVT1"],
                ),
            ),
            Named(
                Method {
                    name: NameString["MTH1"],
                    flags: MethodFlags {
                        arg_count: 0,
                        serialized: false,
                        sync_level: 0,
                    },
                    body: [],
                },
            ),
            Named(
                Mutex {
                    name: NameString["MTX1"],
                    sync_level: 0,
                },
            ),
            Named(
                PowerResource {
                    name: NameString["PWR1"],
                    system_level: 2,
                    resource_order: 99,
                    body: [],
                },
            ),
            Named(
                Processor {
                    name: NameString["PRC1"],
                    id: 1,
                    register_block_addr: 291,
                    register_block_length: 10,
                    body: [],
                },
            ),
            Named(
                ThermalZone {
                    name: NameString["THZ1"],
                    body: [],
                },
            ),
            Modifier(
                Alias {
                    source: NameString["MTH1"],
                    alias: NameString["MTH2"],
                },
            ),
            Modifier(
                Scope(
                    NameString["DEV1"],
                    [
                        Modifier(
                            Name(
                                NameString["FOO_"],
                                Data(
                                    Data(
                                        String(
                                            "bar",
                                        ),
                                    ),
                                ),
                            ),
                        ),
                    ],
                ),
            ),
        ],
    },
)
